torchrl/objectives/sac.py

```diff
@@ -1060,159 +1058,145 @@ def in_keys(self, values):
+            # get probs and log probs for actions computed from "next"
+            with set_exploration_type(ExplorationType.RANDOM):
+                next_dist = self.actor_network.get_dist(
+                    next_tensordict, params=actor_params
+                )
+                next_prob = next_dist.probs
+                next_log_prob = torch.log(torch.where(next_prob == 0, 1e-8, next_prob))
 
-        pred_val_1 = (
-            state_action_value_qvalue[0].gather(-1, actions.unsqueeze(-1)).unsqueeze(0)
+            # get q-values for all actions
+            next_tensordict_expand = self._vmap_qnetworkN0(next_tensordict, qval_params)
+            next_action_value = next_tensordict_expand.get(
+                self.tensor_keys.action_value
+            )
+
+            # like in continuous SAC, we take the minimum of the value ensemble and subtract the entropy term
+            next_state_value = next_action_value.min(0)[0] - _alpha * next_log_prob
+            # unlike in continuous SAC, we can compute the exact expectation over all discrete actions
+            next_state_value = (next_prob * next_state_value).sum(-1).unsqueeze(-1)
+
+            tensordict.set(
+                ("next", self.value_estimator.tensor_keys.value), next_state_value
+            )
+            target_value = self.value_estimator.value_estimate(tensordict).squeeze(-1)
+            return target_value
+
+    def _loss_value(self, tensordict: TensorDictBase) -> Tuple[Tensor, Tensor]:
+        target_value = self._get_target_value(
+            tensordict,
+            self._alpha,
+            self.actor_network_params,
+            self.target_qvalue_network_params,
         )
-        pred_val_2 = (
-            state_action_value_qvalue[1].gather(-1, actions.unsqueeze(-1)).unsqueeze(0)
+        tensordict_expand = self._vmap_qnetworkN0(
+            tensordict.select(*self.qvalue_network.in_keys),
+            self.qvalue_network_params,
         )
-        pred_val = torch.cat([pred_val_1, pred_val_2], dim=0).squeeze()
-        td_error = (pred_val - target_value.expand_as(pred_val)).pow(2)
-        loss_qval = (
-            distance_loss(
-                pred_val,
-                target_value.expand_as(pred_val),
-                loss_function=self.loss_function,
+
+        action_value = tensordict_expand.get(self.tensor_keys.action_value)
+        action = tensordict.get(self.tensor_keys.action)
+
+        # TODO this block comes from the dqn loss, we need to swap all thse with a proper
+        #  helper function which selects the value given the action for all discrete spaces
+        if self.action_space == "categorical":
+            if action.shape != action_value.shape:
+                # unsqueeze the action if it lacks on trailing singleton dim
+                action = action.unsqueeze(-1)
+            chosen_action_value = torch.gather(action_value, -1, index=action).squeeze(
```

Comment:
Just noticed a bug here when I was testing in the categorical case. So in my setup, I have 2 qvalue nets. This causes my action_value to have a shape of: `torch.Size([2, 256, 1, 5])`. However, my action is just of the shape `torch.Size([256, 1, 1])`. Note that it has the extra dimension at the end because of the unsqueeze above. However, this causes the `gather` to fail as the `index` and `input` tensors have mismatching dimensions. Using the following code below seemed to work for me:

```python
action_idx = action.expand(torch.Size([action_value.shape[0], *action.shape]))
chosen_action_value = torch.gather(action_value, -1, index=action_idx).squeeze(-1)
```

Reply:
thanks so much for spotting this, phew.
can you tell me if now it works in both one hot and categorical cases for you?

Reply:
Yep, it works for me!